### PR TITLE
Issue #12221: Change checker-framework.groovy approach to support windows OS

### LINKED ISSUE #12221: checker-framework.groovy (differences): false positives when running checker-framework-nullness-optional-interning
When running `groovy ./.ci/checker-framework.groovy checker-framework-nullness-optional-interning` on a clean master, I get:
````
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  12:34 min
[INFO] Finished at: 2022-09-20T20:47:21-04:00
[INFO] ------------------------------------------------------------------------
New surviving error(s) found:

File Name: "checkstyle\src\main\java\com\puppycrawl\tools\checkstyle\Checker.java"
Specifier: "argument"
Message: "incompatible argument for parameter moduleId of Violation."
Line Contents: "new String[] {ioe.getMessage()}, null, getClass(), null));"
Line Number: 342
Details: found   : null (NullType)
....
````

I get 308 new errors and 643 unnecessary errors.

I a running on Windows.

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #12221

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
This is passing in CI but always fails in my local:
```
Caught: java.io.FileNotFoundException: [WARNING] /home/vyom/IdeaProjects/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java (No such file or directory)
java.io.FileNotFoundException: [WARNING] /home/vyom/IdeaProjects/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java (No such file or directory)
```
Only happens when I run the script, but if I pass in the output of the script from a file, it works fine.

File: .ci/checker-framework.groovy [hunk 1]
```diff
@@ -113,27 +113,43 @@ private static int checkCheckerFrameworkReport(final String profile) {
...
+    while ((lineFromReader = reader.readLine()) != null) {
+        println(lineFromReader)
+        checkerFrameworkLines.add(lineFromReader)
     }
     process.waitFor()
+
+    for (int index = checkerFrameworkLines.size() - 1; index >= 0; --index) {
+        final String curLine = checkerFrameworkLines.get(index)
+        if (curLine.startsWith('[ERROR]') || curLine.startsWith('[INFO]')) {
+            checkerFrameworkLines.remove(index)
```
Comment (romani) [inline .ci/checker-framework.groovy:131]: We removing elements from same collection as we iterate, and recount index - it is a problem, you can miss items.
Better to have new collection for result.

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by rnveach** [comment]
I cloned this branch and ran it on my local and I am still getting a ton of differences.

I posted my run at: http://rveach.no-ip.org:81/checkstyle/regression/nullness-optional-interning-windows-output-PR12366-2022-11-4.txt

Some things I noticed:
* forward slash versus backward slash
* second lines missing (Example: found and required but only tracked the found)
* A few instances of what looks like a possible order issue (Example: 1, 3, 2 instead of 1, 2, 3)

File: .ci/checker-framework.groovy (hunk 1 above)
Comment (nrmancuso) [reply .ci/checker-framework.groovy:131]: Yes, @Vyom-Yadav why add lines to collection just to be removed anyway and iterate through collection twice? Above, we are already iterating over lines, why not add only `[WARNING]` lines to collection after being trimmed?

File: .ci/checker-framework.groovy [hunk 2]
```diff
@@ -113,27 +113,43 @@ private static int checkCheckerFrameworkReport(final String profile) {
...
+        }
+        else {
+            checkerFrameworkLines.set(index, curLine.replace('[WARNING]', '').trim())
+        }
+    }
     final List<List<String>> checkerFrameworkErrors = new ArrayList<>()
     for (int index = 0; index < checkerFrameworkLines.size(); index++) {
         final String line = checkerFrameworkLines.get(index)
-        if (line.startsWith('[WARNING]')) {
+        final String fileMatchingPattern = '.*src[\\\\/].*\\.java.*'
```
Comment (nrmancuso) [inline .ci/checker-framework.groovy:140]: Why do we need dot and kleene star at the end of the regexp? Can we do: `'.*src[\\\\/].*\\.java$`?

File: .ci/checker-framework.groovy [hunk 3]
```diff
@@ -113,27 +113,43 @@ private static int checkCheckerFrameworkReport(final String profile) {
 private static List<List<String>> getCheckerFrameworkErrors(final String profile) {
     final List<String> checkerFrameworkLines = new ArrayList<>()
     final String command = "mvn -e --no-transfer-progress clean compile -P${profile}"
-    final Process process = getOsSpecificCmd(command).execute()
-    process.in.eachLine { final line ->
-        checkerFrameworkLines.add(line)
-        println(line)
+    final ProcessBuilder processBuilder = new ProcessBuilder(getOsSpecificCmd(command).split(' '))
+    processBuilder.redirectErrorStream(true)
```
Comment (nrmancuso) [inline .ci/checker-framework.groovy:117]: Please add comment here about why we do this.

File: .ci/checker-framework.groovy [hunk 4]
```diff
@@ -113,27 +113,43 @@ private static int checkCheckerFrameworkReport(final String profile) {
...
+    final Process process = processBuilder.start()
+
+    final BufferedReader reader = new BufferedReader(new InputStreamReader(process.inputStream))
+    String lineFromReader
+    while ((lineFromReader = reader.readLine()) != null) {
```
Comment (nrmancuso) [inline .ci/checker-framework.groovy:122]: As we do for java, let's not use assignment in control variable

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Items:

File: .ci/checker-framework.groovy (hunk 1 above)
Comment (Vyom-Yadav) [reply .ci/checker-framework.groovy:131]: > We removing elements from same collection as we iterate, and recount index - it is a problem, you can miss items.

That's why I am iterating in reverse order, already examined items won't be affected and the upcoming items index won't change.

Comment (Vyom-Yadav) [reply .ci/checker-framework.groovy:131]: > Yes, @Vyom-Yadav why add lines to collection just to be removed anyway and iterate through collection twice? Above, we are already iterating over lines, why not add only `[WARNING]` lines to collection after being trimmed?

I initially did that, but I was facing https://github.com/checkstyle/checkstyle/pull/12366#issuecomment-1303009194, so I tired this one to see if it makes any difference, reverting back to filtering in a single iteration appraoch.

File: .ci/checker-framework.groovy (hunk 4 above)
Comment (Vyom-Yadav) [reply .ci/checker-framework.groovy]: Done.

File: .ci/checker-framework.groovy (hunk 1 above)
Comment (Vyom-Yadav) [reply .ci/checker-framework.groovy]: Done.

**Comment by Vyom-Yadav** [comment]
@rnveach @romani @nrmancuso I don't think it is possible to support windows without fixing the issue with maven, from the logs rnveach has posted (same logs in my local):

```
M:\checkstyleWorkspace\checkstyle\src\main\java\com\puppycrawl\tools\checkstyle\Checker.java:[342,20] [argument] incompatible argument for parameter args of Violation.
  found   : @Initialized @Nullable String @Initialized @NonNull []
[WARNING] M:\checkstyleWorkspace\checkstyle\src\main\java\com\puppycrawl\tools\checkstyle\Checker.java:[342,53] [argument] incompatible argument for parameter moduleId of Violation.
  found   : null (NullType)
[WARNING] 
  required: @Initialized @NonNull String
```
The corresponding violation is:
```xml
  <checkerFrameworkError unstable="false">
    <fileName>src\main\java\com\puppycrawl\tools\checkstyle\Checker.java</fileName>
    <specifier>argument</specifier>
    <message>incompatible argument for parameter args of Violation.</message>
    <lineContent>new String[] {ioe.getMessage()}, null, getClass(), null));</lineContent>
    <details>
      found   : @Initialized @Nullable String @Initialized @NonNull []
    </details>
  </checkerFrameworkError>
```
this should have been:
```xml
  <checkerFrameworkError unstable="false">
    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/Checker.java</fileName>
    <specifier>argument</specifier>
    <message>incompatible argument for parameter args of Violation.</message>
    <lineContent>new String[] {ioe.getMessage()}, null, getClass(), null));</lineContent>
    <details>
      found   : @Initialized @Nullable String @Initialized @NonNull []
      required: @Initialized @NonNull Object @Initialized @NonNull []
    </details>
  </checkerFrameworkError>
```
**Now this isn't related to logging or the order of logs.** The line ` required: @Initialized @NonNull Object @Initialized @NonNull []` isn't present anywhere in the checker-framework logs.

**Comment by rnveach** [comment]
@Vyom-Yadav This PR is still listed as a draft. I don't feel there is a point in reviewing until even you think it is ready and out of draft, but in the meantime I will point out some things.

> Resolves https://github.com/checkstyle/checkstyle/issues/12221

This doesn't resolve the issue and it looks like it will take a few more fixes to resolve this.

If you feel this PR can be stand alone and continue on, then please finish it. However, we can no longer continue under this SUPER issue as there is too much chatter and too much chaos. Please create a new issue for just what this PR is doing and document in the first post what is specifically being done in general terms. If there is more than 2 items being covered by this PR, then we need to split it up more, where possible. I will close the SUPER issue if we don't need a main base of misc. chatter.

We will need new, separate issues for everything that is wrong with Windows run. Please create one issue for each of what is left and do only 1 PR per issue. Don't try to combine and finish everything in 1 go, it is better to go piece by piece on this monster.

The first thing that should be worked on, again in a separate PR, is: `forward slash versus backward slash in file name`. It creates too much noise to properly compare before and after runs on Windows.

> The line required: @Initialized @NonNull Object @Initialized @NonNull [] isn't present anywhere in the checker-framework logs.

I believe I agree, but please double check its not an ordering issue as I pointed above, and the output is truly no where connected to `Checker.java`.

If this is the case, this will be its own issue and please report this issue to Checker as it is their problem. We will continue on with other fixes we can make in the meantime.

Also please report a separate issue to Checker that we would like a way for them to create a report file so we can completely drop maven output parsing. It is far too fragile and time consuming to continue this long term.

**Comment by nrmancuso** [comment]
I recommend us to:

1. Create new issue + task list with details from https://github.com/checkstyle/checkstyle/pull/12366#issuecomment-1304585080 and https://github.com/checkstyle/checkstyle/issues/12221#issuecomment-1312751568
2. Close issue at #12221, now that we see that #12221 is symptom of larger issues
3. Close this PR

It looks like there are a lot of moving parts here, we will be able to move faster on fixes once we have a detailed plan.

**Comment by rnveach** [comment]
@Vyom-Yadav ping. Do you have any timeframe to return to this?

**Comment by Vyom-Yadav** [comment]
> @Vyom-Yadav ping. Do you have any timeframe to return to this?

Yup, I am closing this and will create separate issues for each problem.

**Pull request closed without merging**